### ISSUE TITLE
replace the existing PR and release blocking kube-up based jobs with CAPI jobs

### ISSUE THREAD

**Issue by neolit123** [kubernetes/kubernetes#82532]
had a discussion with @timothysc about this yesterday.

this is a tracking issue for replacing the pending PR and release blocking jobs based on kube-up with Cluster API (CAPI) provider alternatives over time.

ideally:
- the replacement jobs should co-existing in parallel with the kube-up jobs.
- once the replacement jobs are stable the kube-up jobs should be removed.

xref issue for removing the /cluster directory:
https://github.com/kubernetes/kubernetes/issues/78995
which can be considered a parent of this one.

TODO:
- [x] enumerate the current list of PR and release blocking jobs
assigned: @neolit123

see:
https://github.com/kubernetes/kubernetes/issues/82532#issuecomment-536783242
https://github.com/kubernetes/kubernetes/issues/82532#issuecomment-536802097

- [ ] experiment replacing `pull-kubernetes-e2e-gce` with a CAPA (cluster-api AWS) / CAPG (cluster-api GCP) job
assigned: @mariantalla 
ATM CAPA is the most complete provider. 
there was a discussion [here](https://github.com/kubernetes/kubernetes/issues/78995) and during the sig-arch code organization meeting about this being the first step we can take.
ideally we should be using CAPG (GCP provider).

- [ ] estimate if/how can we replace the big SIG scalability jobs using Cluster API,
assigned: @timothysc
@timothysc mentioned that currently there are TODOs there to enable support for large number of nodes in CAPI.

- [ ] replace SIG-Windows kube-up jobs
assigned: TODO
this is one of the trickier items actually, since CAPI does not support Windows.

/sig testing cloud-provider cluster-lifecycle release scalability
/kind feature
/area test
/priority important-longterm
/assign @spiffxp @justinsb @timothysc 
/assign

**Comment by neolit123** [comment]
> experiment replacing pull-kubernetes-e2e-gce with a CAPA (cluster-api AWS) job

i don't have information on the state of CAPA e2e tests ATM; we had AWS account issues there.
@detiber do you know?

**Comment by neolit123** [comment]
i'd appreciate comments on CAPG (GCP) vs CAPA (AWS).
is it a strict requirement to have GCP jobs that provide signal for Google infrastructure or are we OK with having AWS infrastructure replacements? do we care about infrastructure?

**Comment by timothysc** [comment]
There are open PR's to help uplevel GCP provider.  
/cc @vincepri

**Comment by vincepri** [comment]
GCP PR is here https://github.com/kubernetes-sigs/cluster-api-provider-gcp/pull/143

Would love to get some feedback / requirements to see what are the expectations

**Comment by mariantalla** [comment]
I can start work on the second TODO (=experiment replacing pull-kubernetes-e2e-gce with a CAPA (cluster-api AWS) job)

Here's what I was had in mind more specifically:
**Local dev environment setup**
* Get a cluster (perhaps even a `kind` one will work) and IaaS (GCP?) account
* Deploy prow on it
* Run `pull-kubernetes-e2e-gce` and have it run successfully using that local prow instance
* Take notes on 1) flake rate 2) duration (this data can even come from existing runs of the job) 
**Experimentation**
* Find usage of k/k/cluster/<script> in anything that gets called by `pull-kubernetes-e2e-gce`
* For each mention, replace with Cluster API calls
* Document flake rate and duration; compare with values before (anything else)
* Write up and share

As always please let me know 
* what you think, especially if I've said something outrageous 😅
* if you want to divide and conquer, or if someone has work in progress on this bit.

**Comment by dims** [comment]
@mariantalla for `CAPA (cluster-api AWS)` you wil need a AWS IaaS account i believe. @justinsb @detiber may be able to help get access ..

**Comment by dims** [comment]
Update: CAPG now as a presubmit and a periodic job that starts up a cluster and runs the e2e/conformance tests

- https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-gcp#e2e%20tests
- https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api-provider-gcp#pr-e2e

Thanks to @vincepri for the hand holding and for all the changes in CAPG to be able to get these jobs going quickly.

**Comment by neolit123** [comment]
@dims @vincepri 
nice!

**Comment by neolit123** [comment]
so a political question here would be: can we get the CAPG conformance e2e test job as PR non-blocking ("optional")?
https://godoc.org/k8s.io/test-infra/prow/config#Presubmit

**Comment by neolit123** [comment]
> enumerate the current list of PR and release blocking jobs

blocking presubmits:
https://testgrid.k8s.io/presubmits-kubernetes-blocking

the following jobs are out of scope of this work - i.e. we still have to run them:
```
pull-kubernetes-bazel-build
pull-kubernetes-bazel-test
pull-kubernetes-verify
pull-kubernetes-typecheck
pull-kubernetes-dependencies
pull-kubernetes-cross
pull-kubernetes-integration
```

notes on the other jobs:

----

`pull-kubernetes-e2e-gce` is the primary target of this work and we should attempt to replace it first.
it's maintained by sig-gcp (formally), but the reality is that it's maintained by sig-testing and other volunteers:
https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml
it uses a GCE as the infrastructure provider.
a CAPG jobs should replace it by running the same scope:
`--ginkgo.skip=\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]`
should be noted that it also has:
`--runtime-config=batch/v2alpha1=true`
this is actually already the default value in kubetest, but if in the replacement job we need to pass `--runtime-config` to the api-server explicitly.

this is where the entanglement between test-infra's kubetest and k/cluster can be seen:
based on the "gce" provider kubetest defines a set of env variables that utilities in https://github.com/kubernetes/kubernetes/tree/master/cluster/gce (kube-up) make sense of.

----

`pull-kubernetes-node-e2e` is maintained by sig-node.
https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-node/sig-node-presubmit.yaml
it runs kubetest with a [NOOP `node` deployer](https://github.com/kubernetes/test-infra/blob/master/kubetest/node.go) passing a custom image config:
https://github.com/kubernetes/test-infra/blob/master/jobs/e2e_node/image-config.yaml
and focusing on `"\[NodeConformance\]"`, also uses `--provider=gce`
has to be replaced in a similar way to `pull-kubernetes-node-e2e`

----

```
pull-kubernetes-e2e-gce-100-performance
pull-kubernetes-kubemark-e2e-gce-big
```
these are maintained by sig-scalability and will be the trickiest to replace as the cluster-api is not ready for large scale clusters. they have a lot of parameters:
https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-scalability/sig-scalability-presubmit-jobs.yaml
and help from sig scale will be needed instead of us trying to reverse engineer them.

**Comment by neolit123** [comment]
release blocking jobs:
https://k8s-testgrid.appspot.com/sig-release-master-blocking#Summary

`skew-cluster-latest-kubectl-stable1-gce`
https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cli/sig-cli-config.yaml#L225-L257

this job focuses on `Kubectl` tests.
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-master-new-gci-kubectl-skew/1178789345755140096/build-log.txt

it executes the `./cluster/kubectl.sh` wrapper from k/k and makes sure the kubectl from master supports the api-server from latest stable release.

---

`gce-cos-master-default`

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L142-L172

`--ginkgo.skip=\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]`

(includes notes on GCI images)

reading some GCP release notes GCI probably means COS nowadays?? (Container-Optimized OS)
https://github.com/kubernetes/kubernetes/blob/ef2016c2530bccff26adf3ca20309e10f4b5b97d/cluster/gce/config-default.sh#L68-L74

it creates a single node CP, with 4 workers.
it installs a node problem detector for "gci":
https://github.com/kubernetes/kubernetes/blob/ef2016c2530bccff26adf3ca20309e10f4b5b97d/cluster/gce/config-default.sh#L291

`gci` also results in experimental flags for the kubelet:
https://github.com/kubernetes/kubernetes/blob/ef2016c2530bccff26adf3ca20309e10f4b5b97d/cluster/gce/config-test.sh#L221

the `gci` image comes from here:
https://github.com/kubernetes/kubernetes/tree/3a501844219d06b6fd777a911175b823ab146d43/cluster/gce/gci

---

`gce-cos-master-alpha-features`

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L203-L231

this job takes the latest k8s master and test the following features (while skipping some):
`\[Feature:(Audit|PodPreset|ExpandCSIVolumes|ExpandInUseVolumes|RunAsGroup|TTLAfterFinished|NodeLease|VolumeSnapshotDataSource|CSIInlineVolume)\]|Networking --ginkgo.skip=Networking-Performance|IPv6|Feature:Volumes|csi-hostpath-v0

it again uses the GCI image.

also has some extra env. variables:
```
      - --env=KUBE_FEATURE_GATES=AllAlpha=true,CSIMigration=false
      - --env=KUBE_PROXY_DAEMONSET=true
      - --env=ENABLE_POD_PRIORITY=true
```
according to a github search `ENABLE_POD_PRIORITY` handling does not existing at this point, so this flag is a NO-OP.
`KUBE_PROXY_DAEMONSET` runs kube-proxy as a daemon-set instead of a static pod. looks like it's not the default and this variable is all over the place - including the e2e framework.

i can't find `KUBE_FEATURE_GATES` actually propagating to `--feature-gate` flags, but rather  `--runtime-config` values. this one is quite entangled too.

-----

`gce-cos-master-serial`
https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L309-L335

focuses on Serial and Disruptive tests, skips the rest.

-----

`gce-cos-master-slow`

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L337-L363

`--ginkgo.focus=\[Slow\] --ginkgo.skip=\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]`

-----

`gce-device-plugin-gpu-master`

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-cloud-provider/gcp/gpu/gpu-gce.yaml#L12-L44


it contains some config forker related annotations that i don't understand:
https://github.com/kubernetes/test-infra/blob/59bfc5afff526c46b1c57d8f27e70f0026c0d83c/experiment/config-forker/README.md#supported-annotations

has `--check-leaked-resources` is GCE specific, it checks if the cluster is leaking resources:
https://github.com/kubernetes/test-infra/blob/7d6039b5599e1932743bebbe2966b5e0d2aadbb7/kubetest/e2e.go#L443-L450

by feeding `--gcp-project-type=gpu-project`, Boskos receives this value and creates a specific type of project.

the GPU feature is tested:
`--ginkgo.focus=\[Feature:GPUDevicePlugin\] --minStartupPods=8`

-----

`gci-gce-ingress`

maintained by sig-network

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-network/sig-network-misc.yaml#L208-L236

uses a a couple of env vars:
```
      - --env=GCE_ALPHA_FEATURES=NetworkEndpointGroup
      - --env=KUBE_GCE_ENABLE_IP_ALIASES=true
```

`--ginkgo.focus=\[Feature:Ingress\]|\[Feature:NEG\]`

-----

`node-kubelet-master`

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-node/node-kubelet.yaml#L2-L35

looks similar to the node-e2e presubmit from the previous comment.
runs NodeConformance.

-----

`gce-cos-master-scalability-100`

https://github.com/kubernetes/test-infra/blob/e7447a591e1df8eed08221965802a7342ea2bf1a/config/jobs/kubernetes/sig-scalability/sig-scalability-release-blocking-jobs.yaml#L106-L169

 similar / same as `pull-kubernetes-e2e-gce-100-performance` ??

https://github.com/kubernetes/test-infra/blob/e7447a591e1df8eed08221965802a7342ea2bf1a/config/jobs/kubernetes/sig-scalability/sig-scalability-presubmit-jobs.yaml


----

`integration-master`

https://github.com/kubernetes/test-infra/blob/4e35daa5b4c968dafff1e88baa199da6acb4fa73/config/jobs/kubernetes/sig-testing/integration.yaml#L69-L97

` description: "Ends up running: make test-cmd test-integration"`

**Comment by detiber** [comment]
> so a political question here would be: can we get the CAPG conformance e2e test job as PR non-blocking ("optional")? eventually PR-blocking.
https://godoc.org/k8s.io/test-infra/prow/config#Presubmit

I don't think it would be difficult to get it added, however I would rather we define a job that uses a fixed version of CAPI, CABPK, and CAPG before we go down that path, though.

**Comment by kikisdeliveryservice** [comment]
Hi all,

Bug Triage Release Team checking in. Code freeze for the 1.17 release is on November 18. Is this still intended for 1.17? Thanks!

**Comment by neolit123** [comment]
hi @kikisdeliveryservice 
this effort is going to continue in the next cycle(s).
/milestone v1.18

**Comment by smourapina** [comment]
Hello @neolit123!
Bug Triage team here for the 1.18 release. This is a friendly reminder that code freeze is scheduled for 5 March. Is this issue still intended for milestone 1.18, or should it extend to the next one? Thanks in advance!

**Comment by vincepri** [comment]
Probably should be expended to the next one, @detiber thoughts?

**Comment by neolit123** [comment]
hi, yes this work will continue in the next cycle(s).
/milestone v1.19

**Comment by kbruner** [comment]
Hello from the bug triage team! This issue has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting June 25th (about 4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.

As the issue is tagged for 1.19, is it still planned for this release?

**Comment by neolit123** [comment]
hi, yes this work will continue in the next cycle(s).
/milestone v1.20

**Comment by MonzElmasry** [comment]
Hey :wave: Bug Triage here. Wanted to follow up on the status of this issue as we're approaching code freeze on 12.11.2020. This issue is tagged for 1.20, is it still planned for this release?

**Comment by neolit123** [comment]
Moving to .21
/milestone v1.21

**Comment by vincepri** [comment]
@fabriziopandini ptal at this issue

**Comment by kcmartin** [comment]
Hello! I am from the Bug Triage team! This issue has not been updated for a long time, so I'd like to check on the status. The code freeze is starting March 9th, 2021 (less than 1 week from now). We want to ensure that each PR has a chance to be merged before code freeze.

As the Issue is tagged for 1.21, is it still planned for this release?

**Comment by fabriziopandini** [comment]
Moving to .22
/milestone v1.22

**Comment by voigt** [comment]
As we already approached code freeze for 1.22 and this issue is frozen, I'm going to remove the `milestone 1.22` label.
Feel free to assign it to another milestone, once development on this topic gets traction again.

/milestone clear

**Comment by fabriziopandini** [comment]
I'm ok in closing this issue if there are no more actionable item at this stage

**Comment by neolit123** [comment]
closing.
let's reuse and reopen when/if more discussion is needed. also to add - there is a capz pr blocking job nowadays.
